include/wil/cppwinrt_authoring.h

```diff
@@ -0,0 +1,247 @@
+//*********************************************************
+//
+//    Copyright (c) Microsoft. All rights reserved.
+//    This code is licensed under the MIT License.
+//    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+//    ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+//    TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+//    PARTICULAR PURPOSE AND NONINFRINGEMENT.
+//
+//*********************************************************
+
+namespace wil
+{
+#ifndef __WIL_CPPWINRT_AUTHORING_PROPERTIES_INCLUDED
+#define __WIL_CPPWINRT_AUTHORING_PROPERTIES_INCLUDED
+    namespace details
+    {
+        template<typename T>
+        struct single_threaded_property_storage
+        {
+            T m_value{};
+            single_threaded_property_storage() = default;
+            single_threaded_property_storage(const T& value) : m_value(value) {}
+            operator T& () { return m_value; }
+            operator T const& () const { return m_value; }
+            template<typename Q> auto operator=(Q&& q)
+            {
+                m_value = wistd::forward<Q>(q);
+                return *this;
+            }
+        };
+    }
+
+    template <typename T>
+    struct single_threaded_ro_property : std::conditional_t<std::is_scalar_v<T> || std::is_final_v<T>, wil::details::single_threaded_property_storage<T>, T>
+    {
+        single_threaded_ro_property() = default;
+        single_threaded_ro_property(T value) : base_type(std::move(value)) { }
+        
+        const auto& operator()()
+        {
+            return *this;
+        }
+        template<typename Q> auto operator=(Q&& q) = delete;

```

Comment:
Doesn't deleting this nerf the ability of the implementation object to use this type for its own purposes?  This is a "construct only" property, not a "read only" property. I want to write this:

```c++
// IDL
runtimeclass MyType {
    String DisplayName { get; };
    bool TrySetDisplayName(String displayName);
}

// Implementation
struct MyType : MyTypeT<MyType> {
    wil::single_threaded_ro_property<winrt::hstring> DisplayName;
    bool TrySetDisplayName(winrt::hstring const& id) {
        if (NumberOfSecondsIsEven()) {
            DisplayName = id;
            return true;
        } else {
            return false;
        }        
    }
};
```

There's even a test that verifies this is blocked:

```c++
TEST_CASE("CppWinRTAuthoringTests::InStruct", "[property]")
{
    struct TestStruct
    {
        wil::single_threaded_ro_property<int> Prop1{ 42 };
        wil::single_threaded_rw_property<int> Prop2{ 1 };
        wil::single_threaded_ro_property<int> Prop3{ 44 };
    };

    TestStruct test;
    static_assert(!std::is_assignable_v<wil::single_threaded_ro_property<int>, int>, "cannot assign to a readonly property");
```

Reply:
I _think_ you mean "single_threaded_get_only" or "single_threaded_property" ?

Reply:
I don't understand? I want to write that type I specified above - there's a `DisplayName` read only property, in that the property isn't settable by the caller _except_ in calling the _method_ `TrySetDisplayName`.

The IDL is the _interface_ definition - what people can do with the object. It shouldn't dictate how you _implement_ the object.

It's fine if you want to call this `single_threaded_ro_property` so an AST walker can derive the IDL, but don't block the _implementation_ from updating its value.